src/metabase/legacy_mbql/util.cljc

```diff
@@ -800,15 +800,17 @@
 (defn matching-locations
   "Find the forms matching pred, returns a list of tuples of location (as used in get-in) and the match."
   [form pred]
-  (loop [stack [[[] form]], matches []]
+  ;; Surprisingly enough, a list works better as a stack here than a vector.
```

Comment:
I'm surprised that this is surprising 😄 

Not sure how big this can get, but if you wanted to save some more memory and go mutable `ArrayDeque` would probably work quite well.

Reply:
In general, an array-based queue when amortized beats a LL-based queue, but that is if the array-based one is mutable. Here, since both are immutable, I guess it's not really "surprising".

I wouldn't suggest complicating this function further, it's already not that significant impact-wise to add more WTF factor there.